bam_sort.c

```diff
@@ -51,16 +51,85 @@ DEALINGS IN THE SOFTWARE.  */
 #include "sam_opts.h"
 #include "samtools.h"
 #include "bedidx.h"
+#include "bam.h"
+
+
+// Struct which contains the sorting key for TemplateCoordinate sort.
+typedef struct {
+    int tid1;
+    int tid2;
+    hts_pos_t pos1;
+    hts_pos_t pos2;
+    bool neg1;
+    bool neg2;
+    const char *library;
+    char *mid;
+    char *name;
+    bool is_upper_of_pair;
+} template_coordinate_key_t;
+
+// Struct to store fixed buffers of template coordinate keys
+typedef struct {
+  size_t n; // the # of buffers stored
+  size_t m; // the # of buffers allocated
+  size_t buffer_size; // # the fixed size of each buffer
+  template_coordinate_key_t **buffers; // the list of buffers
+} template_coordinate_keys_t;
+
+// Gets the idx'th key; does not OOB checking
+static template_coordinate_key_t* template_coordinate_keys_get(template_coordinate_keys_t *keys, size_t idx) {
+    size_t buffer_idx = idx / keys->buffer_size; // the index of the buffer to retrieve in buffer
+    size_t buffer_offset = idx % keys->buffer_size; // the offset into the given buffer to retrieve
+    return &keys->buffers[buffer_idx][buffer_offset];
+
+
+}
+
+// Rellocates the buffers to hold at least max_k entries
+static int template_coordinate_keys_realloc(template_coordinate_keys_t *keys, int max_k) {
+    template_coordinate_key_t **new_buffers = NULL;
+    size_t new_m = 1 + (max_k / keys->buffer_size); // new # of buffers needed
+    assert(new_m * keys->buffer_size >= max_k);
+    if ((new_buffers = malloc(new_m * sizeof(template_coordinate_key_t*))) == NULL) {
+        print_error("sort", "couldn't allocate memory for template coordinate key buffers");
+        return -1;
+    }
+    // copy over existing buffers
+    int j;
+    for (j = 0; j < keys->n; ++j) {
+        new_buffers[j] = keys->buffers[j];
+    }
```

Comment:
This looks a bit more complicated than strictly necessary - there's nothing wrong with calling `realloc()` on `keys->buffers`.